code/parachain/runtime/common/src/dex.rs

```diff
@@ -0,0 +1,262 @@
+use crate::{cosmwasm::*, prelude::*};
+use composable_traits::dex::*;
+use cosmwasm_std::{to_binary, Coin, QueryResponse, Response, Uint128};
+use primitives::currency::CurrencyId;
+use sp_runtime::traits::Convert;
+use sp_std::marker::PhantomData;
+
+pub struct DexPrecompile<Balance, AccountId, Dex>(PhantomData<(Balance, AccountId, Dex)>);
+
+pub enum CosmwasmPrecompileError {
+	AssetConversion,
+	AccountConversion,
+	DispatchError,
+	Serde,
+}
+
+impl<Dex, Balance, AccountId> DexPrecompile<Balance, AccountId, Dex>
+where
+	Dex:
+		Amm<AssetId = CurrencyId, Balance = Balance, AccountId = AccountId, PoolId = crate::PoolId>,
+	Balance: From<Uint128> + Into<u128>,
+	AccountId: sp_std::convert::From<sp_runtime::AccountId32>,
+{
+	fn to_substrate(
+		in_asset: &Coin,
+	) -> Result<AssetAmount<CurrencyId, Balance>, CosmwasmPrecompileError> {
+		let in_asset_id = CosmwasmToSubstrateAssetId::convert(in_asset.denom.clone())
+			.map_err(|_| CosmwasmPrecompileError::AssetConversion)?;
+		let in_amount: Balance = in_asset.amount.into();
+		Ok(AssetAmount::new(in_asset_id, in_amount))
+	}
+
+	fn to_cw(amount: AssetAmount<CurrencyId, Balance>) -> Coin {
+		Coin {
+			denom: CosmwasmToSubstrateAssetId::convert(amount.asset_id),
+			amount: amount.amount.into().into(),
+		}
+	}
+
+	pub fn query(sender: &str, msg: QueryMsg) -> Result<QueryResponse, CosmwasmPrecompileError> {
+		match msg {
+			QueryMsg::Assets { pool_id } => {
+				let assets = Dex::assets(pool_id.into())
+					.map_err(|_| CosmwasmPrecompileError::DispatchError)?;
+				let assets = assets
+					.into_iter()
+					.map(|(k, v)| {
+						let fraction = ((v.deconstruct() as u64).into(), 1_000_000_u64.into());
+						let denom = CosmwasmToSubstrateAssetId::convert(k);
+						(denom, fraction)
+					})
```

Comment:
OK, got it.  Getting better variable names would make this clearer.
Also, prefer `u64::from(x)` to `x as u64` since the former will fail
at compilation time if the conversion is lossy.

```suggestion
					.map(|(asset_id, amount_permill)| {
						let numerator = u64::from(amount_permill.deconstruct);
						let fraction = (numerator.into(), 1_000_000_u64.into());
						let denom = CosmwasmToSubstrateAssetId::convert(asset_id);
						(denom, fraction)
					})
```